src/GitHub/Query.hs

```diff
@@ -159,33 +197,45 @@ callGitHubRaw (GitHubToken token) text = do
 
     -- calling the API
     response <- httpLbs request manager
+    let responseCode = statusCode $ responseStatus response
+    let jsonBody     = responseBody response
+
+    let mkError :: Maybe String -> Maybe String -> [GitHubErrorDetails] -> GitHubError
+        mkError gitHubErrorJsonRes gitHubErrorJsonErr gitHubErrorDetails = GitHubError
+            { gitHubErrorResponseCode = responseCode
+            , gitHubErrorQuery        = query
+            , gitHubErrorJsonBody     = jsonBody
+            , ..
+            }
 
-    case statusCode $ responseStatus response of
-        200 -> case eitherDecode @(GitHubDataResponse a) (responseBody response) of
-            Left err ->
-                throwIO $ userError $ "Error decoding JSON response: " <> err
-            Right (GitHubDataResponse res) ->
-                pure res
-        code -> throwIO $ userError $ "Non-200 response code: " <> show code
-
-{- | Local wrapper to parse objects inside the @data@ field. GraphQL
-API returns JSON of the following shape.
+    pure $ case responseCode of
+        200 -> case decodeResult jsonBody of
+            Right res   -> Right res
+            Left resErr -> case decodeErrors jsonBody of
+                Right errs      -> Left $ mkError (Just resErr) Nothing errs
+                Left detailsErr -> Left $ mkError (Just resErr) (Just detailsErr) []
 
-@
-{
-    "data": { ... actual result ... }
-}
-@
-
-This wrapper handles extra layer of @data@.
+        _code -> Left $ mkError Nothing Nothing []
+  where
+    decodeResult :: LBS.ByteString -> Either String a
+    decodeResult = fmap unNested . eitherDecode @(Nested '[ "data" ] a)
+
+    decodeErrors :: LBS.ByteString -> Either String [GitHubErrorDetails]
+    decodeErrors = fmap unNested . eitherDecode @(Nested '[ "errors" ] [GitHubErrorDetails])
+
+data GitHubError = GitHubError
+    { gitHubErrorResponseCode :: Int
+    , gitHubErrorQuery        :: Text
+    , gitHubErrorJsonBody     :: LBS.ByteString
+    , gitHubErrorJsonRes      :: Maybe String
+    , gitHubErrorJsonErr      :: Maybe String
+    , gitHubErrorDetails      :: [GitHubErrorDetails]
+    } deriving stock (Show, Eq)
+
+{- | Helper function to work with GraphQL Queries that return 'Either'.
 -}
-newtype GitHubDataResponse a = GitHubDataResponse a
-
-instance (Typeable a, FromJSON a) => FromJSON (GitHubDataResponse a) where
-    parseJSON = withObject ("GitHubDataResponse " ++ typeName @a) $ \o -> do
-        val <- o .: "data"
-        pure $ GitHubDataResponse val
-
--- copy-pasted from @relude@
-typeName :: forall a. Typeable a => String
-typeName = show (typeRep @a)
+unNest
+    :: forall keys a
+    .  IO (Either GitHubError (Nested keys a))
+    -> IO (Either GitHubError a)
+unNest = fmap (second unNested)
```

Comment:
Maybe Errors could be in a separate module? Or you think that its interface won't change a lot?

Reply:
That's a great idea!